rfcs/20230503-profiling-api/README.md

```diff
@@ -0,0 +1,196 @@
+to query the profiling data for the primitive
+
+In order to support profiling for multiple streams each stream will have its own
+profiler, this is why the profiling API will have a stream as a parameter.
+
+A typical workflow looks as follows:
+* Enable profiling
+* Create a oneDNN stream with a queue that was created with enabled profiling
+capabilities
+* Reset the profiler's state before executing the primitive
+* Execute the primitive and explicitly wait for it to complete
+* Use the corresponding API to query profiling data
+* Reset the profiler's state
+* Disable profiling
+
+### API
+
+The API for enabling and disabling profiling:
+
+#### C
+```c
+/// Enable profiling capabilities.
+///
+/// @param stream Stream to be used for profiling.
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_enable_profiling(dnnl_stream_t stream);
+
+/// Disable profiling capabilities.
+///
+/// @param stream Stream that is used for profiling
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_disable_profiling(dnnl_stream_t stream);
+```
+
+#### C++
+```cpp
+/// @copydoc dnnl_enable_profiling()
+status enable_profiling(stream &stream);
+
+/// @copydoc dnnl_disable_profiling()
+status disable_profiling(stream &stream);
+```
+
+The API for resetting the profiler's state:
+
+#### C
+```c
+/// Resets a profiler's state.
+///
+/// @param stream Stream associated with the profiler.
+///
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_reset_profiling(dnnl_stream_t stream);
+```
+
+#### C++
+```cpp
+/// @copydoc dnnl_reset_profiling()
+status reset_profiling(stream &stream);
+```
+
+The API to query profiling data.
+
+#### C
+
+```c
+/// Profiling data kind.
+typedef enum {
+    /// Undefined profiling data kind.
+    dnnl_profiling_data_kind_undef = 0,
+    /// Data kind to query an execution time in nanoseconds.
+    dnnl_profiling_data_kind_time,
+} dnnl_profiling_data_kind_t;
+
+/// Profiling mode.
+typedef enum {
+    /// Undefined profiling mode.
+    dnnl_profiling_mode_undef = 0,
+    /// The profiling data for each kernel will be added up. For example,
+    /// if the profiling data kind is time and there are n kernels then the
+    /// result will be timeK0 + timeK1 + timeKn-1.
+    dnnl_profiling_mode_sum,
+} dnnl_profiling_mode_t;
+
+
+/// Queries profiling data.
+///
+/// @param stream Stream that was used for executing a primitive that
+/// is being profiled.
+/// @param data_kind Profiling data kind to query.
+/// @param mode Profiling mode.
+/// @param data Profiling data.
+///
+/// @returns #dnnl_success on success and a status describing the error
+///     otherwise.
+dnnl_status_t dnnl_query_profiling_data(dnnl_stream_t stream,
+        dnnl_profiling_data_kind_t data_kind,
+        dnnl_profiling_mode_t mode, uint64_t *data);
```

Comment:
It would be nice to migrate benchdnn to this public API as now it relies on the internal API:

https://github.com/oneapi-src/oneDNN/blob/dc635ac7ebcd7dac2c8e09132bdd84386ed449ec/tests/benchdnn/dnnl_common.cpp#L412-L417

benchdnn needs data per each primitive execution, in general it queries time and frequency for all profiled primitive execution (one entry per all kernels enqueued in the scope of `primitive.execute()`).

For that we could introduce additional values in `dnnl_profiling_data_kind_t` (even internal, e.g. `dnnl_profiling_data_kind_time_per_execution` and `dnnl_profiling_data_kind_freq_per_execution`) and a query mechanism (with `data_size` parameter) to retrieve time/frequency value per each primitive execution.

Reply:
Yeah, I'm planning to migrate benchdnn. I also plan to preserve the current behavior, i.e. the data accumulates until the profiler is reset. With that we don't seem to need the `per_execution` kinds. We will need `dnnl_profiling_data_kind_freq` though.

Reply:
benchdnn needs information about the minimum and average time across primitive runs. The average time can be computed based on the API from this RFC but for minimum time we either need another kind/mode or allow to query an array of entries (entry per primitive execution).

Reply:
Ok, it seems that we can repurpose the mode enum for this, which looks like a better option to me. Returning an array of entries would look a bit weird because each entry is a kernel and a primitive can have multiple kernels.

Reply:
I mistakenly assumed that we return data for each kernel execution and not primitive execution. After some thinking I realized that returning an array would be more transparent.